implementations/rust/ockam/ockam_api/src/cloud/project.rs

```diff
@@ -301,14 +305,26 @@ mod node {
             trace!(target: TARGET, %space_id, project_name = %req_body.name, "creating project");
 
             let req_builder = Request::post(format!("/v0/{space_id}")).body(req_body);
+            let cli_state = cli_state::CliState::new()?;
+
+            let default_vault = cli_state.vaults.default()?.config.get().await?;
+            let ident = Some(
+                cli_state
+                    .identities
+                    .default()?
+                    .config
+                    .get(ctx, &default_vault)
+                    .await?,
+            );
```

Comment:
This part is a little bit tricky to understand and it's not doing what we need to. At this point, the `req_body` will have an optional identity name:
- If `None`, we want to use the node's default identity
- If `Some`, we want to load that identity using the node's vault and if it fails, fallback to the default vault and rise an error if the identity can't be loaded then

Now, to access the node's vault/identity you can do the following:
```rust
let inner = let inner = self.get().read().await;
let vault = inner.vault()?;
let identity = inner.identity()?;
```

Let me know if you have any questions on this! 👍

Reply:
Gotcha, I might have some time to do these tomorrow night. I'll beep you back if I run into any trouble!

Reply:
I've tried my hand at it but I'm not quite understanding a few steps:
1. Just to confirm - if `None` is a match for `req_body`'s identity name then I can use the `default_vault` approach with lines 310~318, right?

2. From this tidbit: 
```rust
let inner = let inner = self.get().read().await;
let vault = inner.vault()?;
let identity = inner.identity()?;
```
When you say load that identity from `red_body` into the node's vault, do you mean just using `inner.identity()?` I'm confused as to how you would associate that identity with a vault, having just an identity name. 

3. If it is a matter of running `inner.identity()`, that function seems to return a borrowed value, rendering me unable to pass it onto the  `request_controller` afterwards. I couldn't find a clone method implemented for getting that value, should that be implemented?

Reply:
1. No, that's the key point here. We want to use the node's default vault, not the global default vault.
2. Check the code below
3. There is a method for this: `async_try_clone`

Since this is a little tricky, here is a possible solution for this:

```rust
let cli_state = cli_state::CliState::new()?;
let inner = self.get().read().await;
let identity = match &req_body.identity_name {
    Some(idt_name) => {
        let idt_cfg = cli_state.identities.get(idt_name)?.config;
        match idt_cfg.get(ctx, inner.vault()?).await {
            Ok(idt) => idt,
            Err(_) => {
                let vault_cfg = cli_state.vaults.default()?.config;
                idt_cfg.get(ctx, &vault_cfg.get().await?).await?
            }
        }
    }
    None => inner.identity()?.async_try_clone().await?,
};
```

Reply:
One final note: once you have this in place, you should modify the `request_controller` and the `controller_secure_channel` functions to actually use this identity.

You can remove [this block](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_api/src/cloud/mod.rs#L152-L155) and pass the identity to replace [this one](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_api/src/cloud/mod.rs#L121).

Reply:
Thanks a lot for the previous snippet, that was much simpler solution than a series of `if let`s that I was employing at the time and it did help out a lot.

Regarding this last note:
> One final note: once you have this in place, you should modify the `request_controller` and the `controller_secure_channel` functions to actually use this identity.
> 
> You can remove [this block](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_api/src/cloud/mod.rs#L152-L155) and pass the identity to replace [this one](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_api/src/cloud/mod.rs#L121).

After performing those steps, around 24 calls to the `request_controller()` function that were using `None` as a default `ident` value had to be modified to take in an `Identity<Vault>`.

The only working way I found to get an `Identity<Vault>` in those calls was using the initial tip you provided, like such (as seen in [this snippet](https://github.com/build-trust/ockam/pull/3977/files#diff-ac92631eb44c2b0ea45d6c7ef8bb4dec67639f735f6a46d708a61b483125d2feR90-R92)):

```rust 
let inner = let inner = self.get().read().await;
let ident = inner.identity()?.async_try_clone().await;
drop(inner); // Without this drop the compiler would implode complaining of multiple borrows from inner
```

I'm not sure if this is the intended behavior or if we need that same approach in `CreateProject` for every one of these calls. Let me know if this works!

Reply:
> After performing those steps, around 24 calls to the request_controller() function that were using None as a default ident value had to be modified to take in an Identity<Vault>.

Yeah, that's right, I have to create more issues to modify the other endpoints so we have the same behavior as the `create_project` method.

Regarding the use of `drop(inner)`, just in case this is useful, it happened because you were trying to hold a reference to the lock (`inner`) after the function returned its result. The "clean" way to do that is to create an inner scope to work with the lock, so it can be automatically dropped when we get out of that scope:

```rust
let ident = {
    let inner = self.get().read().await;
    inner.identity()?.async_try_clone().await?
};
```

Let me know if you want to do that change, it can be a helpful exercise to better understand the borrow checker. If not, I can do it myself!

The PR looks great now, I'll merge it as soon as we apply the `drop` changes.

Reply:
Great to know! I'll apply those ASAP. Thanks again @adrianbenavides!